### DISCUSSION TITLE
Question about the pressure in THM

### DISCUSSION THREAD

**Discussion by Ethan-xj** [idaholab/moose#24529] [Q&A Modules: General]
@joshuahansel Hi josh, I find that the lateral velocity calculated now is smaller than the common subchannel program. And in the lateral momentum equation, the lateral pressure gradient has a certain effect. So I'm wondering if it's because the solving method of pressure in THM is different. In my understanding, pressure is always calculated as a nonlinear variable. But in THM, pressure is calculated as a material property. Do you think this will have any influence?
Btw I noticed that there was only a area gradient term in axial momentum equation. Isn't there should be a pressure gradient term in the axial momentum equation?

Ethan

**Comment by joshuahansel** [comment]
Sorry for the delay. What is "the common subchannel program"? You say "pressure is always calculated as a nonlinear variable". I don't know what you mean by "always" - it's definitely a choice of the flow formulation. Maybe you mean "often" or "usually". There are formulations where pressure is a solution variable and ones where it's not. The one in THM, like you said, does not have pressure as a solution variable - it has a compressible formulation with {rhoA, rhouA, rhoEA} as solution variables. In theory, it should not matter which set of variables you choose to be the solution variables if you solve the nonlinear system completely. It should only matter what the governing equations are. Maybe "the common subchannel program" uses an incompressible flow formulation, and yes, this is a potential cause of discrepancy in the solutions. As far as variable vs. material property, this should not matter in our case, since THM's variables are piecewise constant on each element. If they were piecewise linear for instance, there would be a difference between values of variables vs. material properties at a given quadrature point.

Yes, there is an area gradient term on the momentum equation. See `ADOneD3EqnMomentumAreaGradient`.

**Reply by Ethan-xj** [reply]
> Maybe you mean "often" or "usually"

Yes, it's my fault. "often" or "usually" is more suitable.

> Maybe "the common subchannel program" uses an incompressible flow formulation, and yes, this is a potential cause of discrepancy in the solutions.

OK, I got it.

> Yes, there is an area gradient term on the momentum equation. See ADOneD3EqnMomentumAreaGradient.

@joshuahansel I didn't express clearly. There should be a pressure gradient term like d(PA)/dx in the momentum equation right? And d(PA)/dx = PdA/dx + AdP/dx. I can only find the PdA/dx (`ADOneD3EqnMomentumAreaGradient`). Shouldn't there be a AdP/dx?

Ethan

**Reply by joshuahansel** [reply]
The d(pA)/dx term is added in `ADNumericalFlux3EqnDGKernel`. So you actually end up with d(pA)/dx - p dA/dx.

**Reply by Ethan-xj** [reply]
@joshuahansel All right, there's no problem with the governing equation.
Btw, can you give me any advice if I want to build a incompressible flow model in THM? I want to find out the reason why there is an error with the temperature.

Ethan

**Reply by Ethan-xj** [reply]
Rebuilding a model seems complicated, maybe is there any way to only replace the way to calculate the material property?

**Reply by joshuahansel** [reply]
Making an incompressible flow model will probably be a lot of work. I think you would just need to write an entirely new component. If you don't want to change the flow model and just want to change some material properties, you need to override whatever method in `FlowModelSinglePhase` that is creating the material with that property. It might mean you end up duplicating a lot of code in that method to create the other objects.

**Reply by Ethan-xj** [reply]
> Making an incompressible flow model will probably be a lot of work.

@joshuahansel I agree with that. First I plan to create a new incompressible flow model. But I realize that if I do that, I need to change all the components which are relative with `FlowChannel1Phase`. But if I make a direct modification on the `FlowModelSinglePhase`, I won't be able to use the compressible flow model. Is there a way that can keep the compressible flow model and build a incompressible flow model at the same time?

Ethan

**Reply by joshuahansel** [reply]
To keep compatibility with all of the other components, your best bet is probably to inherit your new flow model from `FlowModelSinglePhase`. That way, other components still recognize it as the original. Before doing any real work, test this concept: inherit from `FlowModelSinglePhase` but don't change anything. Then in your component, use the new flow model and see if all of the other components complain.

If this incompressible flow model were to be something that ended up being contributed to THM, I'd probably think about a better design for all of this. We'd probably factor some commonality out of the flow model and flow components into base classes that represent all of single-phase flow, not just single-phase *compressible* flow. But I'm assuming you're mainly doing this as troubleshooting.

I hope that you've exhausted other troubleshooting steps first before undertaking the effort: at this point, are you pretty confident that it's usage of a compressible model vs. incompressible one? Nothing wrong with your channel coupling components or heat transfer?

**Reply by Ethan-xj** [reply]
@joshuahansel What you say makes sense. I will continue to test if there's anything wrong with my coupling.
But there still exists problems for the calculation of liquid metal. In `SimpleFluidProperties`, viscosity and thermal conductivity are constant, it can't describe the flow of liquid metal in an assembly correctly. And I think there is a fundamental error if I mix `SimpleFluidProperties` and `LeadBismuthFluidProperties`.So...It's an annoying question.

**Reply by joshuahansel** [reply]
What fluid(s) are you trying to model? Lead bismuth? What is this "mixing" you're referring to?

**Reply by Ethan-xj** [reply]
@joshuahansel liquid_metal like sodium, lead and LBE.
The "mixing" is mentioned in this issue: https://github.com/idaholab/moose/discussions/24197#discussioncomment-5829332

**Reply by GiudGiud** [reply]
I would still keep with SimpleFP if not even something constant until you have something working. Once it's working, then you can transition to an incompressible fluid like lead, and see if the compressible fluid solver can handle that (which I think it cannot).

@Ethan-xj what is your affiliation? There is a code we could use with incompressible fluid flow formulation, it's just export controlled.

**Comment by Ethan-xj** [comment]
@GiudGiud Hello Giud. Thanks for your suggestion and I'm afraid that my affiliation is restricted.
@joshuahansel Hi josh, I think the material property may not be the most crucial problem. I calculate a two-channels benchmark and the calculated values do not agree with the theoretical values. So there's still something wrong with the equations.
Here is the main part of the lateral equation:
```
//* This file is part of the MOOSE framework
//* https://www.mooseframework.org
//*
//* All rights reserved, see COPYRIGHT for full restrictions
//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
//*
//* Licensed under LGPL 2.1, please see LICENSE for details
//* https://www.gnu.org/licenses/lgpl-2.1.html

#include "FLARE_LateralMomentum_Test.h"
#include "FLARE_SolutionVariableUO_A.h"
#include "SinglePhaseFluidProperties.h"

registerMooseObject("ThermalHydraulicsApp", FLARE_LateralMomentum_Test);

InputParameters
FLARE_LateralMomentum_Test::validParams()
{
  InputParameters params = ADKernel::validParams();
  params.addParam<Real>("lk", 0.00242 ,"Length of the turbulence");
  params.addParam<Real>("sk", 0.00142 ,"Width of the gap");
  params.addParam<Real>("kg", 0.05 ,"Drag coefficient of the cross flow");
  params.addRequiredParam<MaterialPropertyName>("rho", "Density property");
  params.addRequiredParam<MaterialPropertyName>("p", "Pressure property");
  params.addRequiredParam<UserObjectName>(
      "q_uo", "The name of the user object that transfers the variables");
  params.addRequiredParam<UserObjectName>("fp", "The name of the user object for fluid properties");
  params.addClassDescription("Computes the lateral velocity between 2 channels.");
  return params;
}

FLARE_LateralMomentum_Test::FLARE_LateralMomentum_Test(const InputParameters & parameters)
  : ADKernel(parameters),
    _rho(getADMaterialProperty<Real>("rho")),
    _p(getADMaterialProperty<Real>("p")),
    _lk(getParam<Real>("lk")),
    _sk(getParam<Real>("sk")),
    _kg(getParam<Real>("kg")),
    _q_uo(getUserObject<FLARE_SolutionVariableUO_A>("q_uo")),
    _fp(getUserObject<SinglePhaseFluidProperties>("fp"))

{
}

ADReal
FLARE_LateralMomentum_Test::computeQpResidual()
{
  ADReal adj_A = _q_uo.getAdjacentA(_current_elem->id());
  ADReal adj_rhoA = _q_uo.getAdjacentRHOA(_current_elem->id());
  ADReal adj_rhouA = _q_uo.getAdjacentRHOUA(_current_elem->id());
  ADReal adj_rhoEA = _q_uo.getAdjacentRHOEA(_current_elem->id());
  ADReal adj_rho = adj_rhoA/adj_A;
  ADReal adj_v = 1.0/adj_rho;
  ADReal adj_e = (adj_rhoEA-0.5*adj_rhouA*adj_rhouA/adj_rhoA)/adj_rhoA;
  ADReal adj_p = _fp.p_from_v_e(adj_v,adj_e);
  return  -_test[_i][_qp]*_sk/_lk*(_p[_qp]-adj_p)+
          _test[_i][_qp]*_kg*std::abs(_u[_qp])*_u[_qp]/(_sk*_lk*(_rho[_qp]+adj_rho));
}
```
And I've already showed the time derivative part and the convective part before. Here is the equation:
![image](https://github.com/idaholab/moose/assets/109257685/48849382-247b-40c2-8dff-1958faf3d1dc)
Did I miss any necessary part of the code?

Ethan

**Reply by joshuahansel** [reply]
Your source terms look fine. I assume your time derivative kernels are fine too. The most likely culprit is your advective term, so I think you'll want to check that one very carefully (verify that you are adding correct values to residuals - don't just look at your code).

**Reply by Ethan-xj** [reply]
@joshuahansel Yes you're right. I find that even if I add the add the `addRDGMooseObjects()` in `addMooseObjects()`, the advective term still do not work. So I output the process parameters to check the code. And I find the problem is because of the `_normals[_qp]`. Honestly, I don't know what the normal vector of 1D mesh should be. But in the case ad_dg_convection.i The mesh is in x direction and the _normals[_qp] is (1,0,0). So since my flow channel is along z direction, I write the code like: `ADReal vdotn = average_vel * _normals[_qp](2)` which is wrong. Because I find that the _normals[_qp] of z-direction-mesh is still (1,0,0). Now I change the code to `ADReal vdotn = average_vel * _normals[_qp](0);` The code works, but I think the` _normals[_qp]` is strange, how should I use it?
And there's another problem I find recently, It's about the other 3 equations. w will be coupled to the mass, momentum and energy equations. For example, in the axial momentum equation:
![image](https://github.com/idaholab/moose/assets/109257685/ef324515-3076-4c9d-909f-394cbfcd8eaa)

In the lateral term w*u, u depends on the direction of the w.
The lateral term code is previously:
```
//* This file is part of the MOOSE framework
//* https://www.mooseframework.org
//*
//* All rights reserved, see COPYRIGHT for full restrictions
//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
//*
//* Licensed under LGPL 2.1, please see LICENSE for details
//* https://www.gnu.org/licenses/lgpl-2.1.html

#include "FLARE_MomentumLateralTerm_S_Test.h"
#include "FLARE_SolutionVariableUO_A.h"

registerMooseObject("ThermalHydraulicsApp", FLARE_MomentumLateralTerm_S_Test);

InputParameters
FLARE_MomentumLateralTerm_S_Test::validParams()
{
  InputParameters params = ADKernel::validParams();
  params.addRequiredCoupledVar("w", "lateral velocity");
  params.addRequiredParam<MaterialPropertyName>("vel", "Velocity property");
  params.addRequiredParam<UserObjectName>(
      "q_uo", "The name of the user object that transfers the variables");
  params.addClassDescription("Computes the lateral velocity between 2 channels.");
  return params;
}

FLARE_MomentumLateralTerm_S_Test::FLARE_MomentumLateralTerm_S_Test(const InputParameters & parameters)
  : ADKernel(parameters),
    _w(adCoupledValue("w")),
    _vel(getADMaterialProperty<Real>("vel")),
    _q_uo(getUserObject<FLARE_SolutionVariableUO_A>("q_uo"))
{
}

ADReal
FLARE_MomentumLateralTerm_S_Test::computeQpResidual()
{
  ADReal adj_rhoA = _q_uo.getAdjacentRHOA(_current_elem->id());
  ADReal adj_rhouA = _q_uo.getAdjacentRHOUA(_current_elem->id());
  ADReal adj_vel = adj_rhouA/adj_rhoA;
  if(_w[_qp] > 0)
  {
    return  _test[_i][_qp]*_w[_qp]*_vel[_qp];
  }
  else
  {
    return  _test[_i][_qp]*_w[_qp]*adj_vel;
  }
}
```
But I realize that maybe I should use the w of last time step to decide the form of u, so I modify the code:
```
//* This file is part of the MOOSE framework
//* https://www.mooseframework.org
//*
//* All rights reserved, see COPYRIGHT for full restrictions
//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
//*
//* Licensed under LGPL 2.1, please see LICENSE for details
//* https://www.gnu.org/licenses/lgpl-2.1.html

#include "FLARE_MomentumLateralTerm_S_Test.h"
#include "FLARE_SolutionVariableUO_A.h"

registerMooseObject("ThermalHydraulicsApp", FLARE_MomentumLateralTerm_S_Test);

InputParameters
FLARE_MomentumLateralTerm_S_Test::validParams()
{
  InputParameters params = ADKernel::validParams();
  params.addRequiredCoupledVar("w", "lateral velocity");
  params.addRequiredParam<MaterialPropertyName>("vel", "Velocity property");
  params.addRequiredParam<UserObjectName>(
      "q_uo", "The name of the user object that transfers the variables");
  params.addClassDescription("Computes the lateral velocity between 2 channels.");
  return params;
}

FLARE_MomentumLateralTerm_S_Test::FLARE_MomentumLateralTerm_S_Test(const InputParameters & parameters)
  : ADKernel(parameters),
    _w(adCoupledValue("w")),
    _w_older(coupledValueOld("w")),
    _vel(getADMaterialProperty<Real>("vel")),
    _q_uo(getUserObject<FLARE_SolutionVariableUO_A>("q_uo"))
{
}

ADReal
FLARE_MomentumLateralTerm_S_Test::computeQpResidual()
{
  ADReal adj_rhoA = _q_uo.getAdjacentRHOA(_current_elem->id());
  ADReal adj_rhouA = _q_uo.getAdjacentRHOUA(_current_elem->id());
  ADReal adj_vel = adj_rhouA/adj_rhoA;
  if(_w_older[_qp] > 0)
  {
    return  _test[_i][_qp]*_w[_qp]*_vel[_qp];
  }
  else
  {
    return  _test[_i][_qp]*_w[_qp]*adj_vel;
  }
}
```
After that, the solution is closer to the theoretical solution. But still not approach the theoretical solution. There should be other errors present.
Btw, The benchmark is here:
[CTF Void Drift Validation Study.pdf](https://github.com/idaholab/moose/files/11732004/CTF.Void.Drift.Validation.Study.pdf)
In the page 3-4, Two-Channel Friction Model Verification.

**Reply by joshuahansel** [reply]
> Honestly, I don't know what the normal vector of 1D mesh should be

What did you set the `orientation` parameter for your component to be? The normal vector is in the context of the DG kernel, so we're on an internal side in your 1D mesh. You're on one element, and you have a neighbor element. The normal points from your element to the neighbor element.

Regarding your lateral velocities, the modification you made might be necessary if you have issues with nonlinear convergence. Otherwise, the original may be ok. One question that comes to mind is: do you use this lateral momentum kernel for both of your channels? If so, then you have a problem because one of your channels needs to flip the sign on the conditional.

**Reply by Ethan-xj** [reply]
@joshuahansel The `orientation` is '0 0 1', here is my input file:
```
T_in = 473.15         # K
# m_dot_in = 1e-4     # kg/s
press = 15.5e6         # Pa

[GlobalParams]
  initial_p = ${press}
  initial_vel = 6.85
  initial_T = ${T_in}

  rdg_slope_reconstruction = full
  closures = simple_closures
  fp = he
[]

[FluidProperties]
  [./he]
#    type = StiffenedGasFluidProperties
#    gamma = 2.35
#    cv = 1816.0
#    q = -1.167e6
#    p_inf = 1.0e9
    type = SimpleFluidProperties
  [../]
[]

[Materials]
  [f_mat]
    type = ADWallFrictionBareRodMaterial
    D_h = D_h
    f_D = f_D
    mu = mu
    rho = rho
    vel = vel
  [../]
[]

[Closures]
  [simple_closures]
#    type = Closures1PhaseSimple
    type = Closures1PhaseNone
  []
[]

[Components]
  [./inlet1]
    type = InletMassFlowRateTemperature1Phase
    input = 'core_chan1:in'
    m_dot = 6
    T = 473.15
  [../]
  [core_chan1]
    type = FlowChannel1Phase
    position = '0 0 0'
    orientation = '0 0 1'
    length = 500
    n_elems = 500
    A = 0.001
    D_h = 0.1333
  [../]
  [outlet1]
    type = Outlet1Phase
    input = 'core_chan1:out'
    p = 15.5e6
  [../]
  [./inlet2]
    type = InletMassFlowRateTemperature1Phase
    input = 'core_chan2:in'
    m_dot = 12
    T = 473.15
  [../]
  [core_chan2]
    type = FlowChannel1Phase
    position = '0.5 0 0'
    orientation = '0 0 1'
    length = 500
    n_elems = 500
    A = 0.002
    D_h = 0.2667
  [../]
  [outlet2]
    type = Outlet1Phase
    input = 'core_chan2:out'
    p = 15.5e6
  [../]
  [Gap]
    type = Gap_Test
    current_flow_channel = core_chan1
    adjacent_flow_channel = core_chan2
    fp = he
  [../]
[]

[Preconditioning]
  [./a]
    type = SMP
    full = true
  [../]
[]

[VectorPostprocessors]
  [./M_SC1]
    type = ElementValueSampler
    variable = rhouA
    sort_by = z
    block = 'core_chan1'
  [../]
  [./M_SC2]
    type = ElementValueSampler
    variable = rhouA
    sort_by = z
    block = 'core_chan2'
  [../]
  [./W_SC1]
    type = ElementValueSampler
    variable = w
    sort_by = z
    block = 'core_chan1'
  [../]
  [./E_SC1]
    type = ElementValueSampler
    variable = rhoEA
    sort_by = z
    block = 'core_chan1'
  [../]
  [./E_SC2]
    type = ElementValueSampler
    variable = rhoEA
    sort_by = z
    block = 'core_chan2'
  [../]
  [./P_SC1]
    type = ElementValueSampler
    variable = p
    sort_by = z
    block = 'core_chan1'
  [../]
  [./P_SC2]
    type = ElementValueSampler
    variable = p
    sort_by = z
    block = 'core_chan2'
  [../]
[]

[Executioner]
  type = Transient
  start_time = 0
  end_time = 100
  dt = 0.1

#  line_search = basic
  solve_type = 'PJFNK'
  l_max_its = 100

  nl_rel_tol = 1e-5
  nl_abs_tol = 1e-5
  nl_max_its = 40

  automatic_scaling = true
  off_diagonals_in_auto_scaling = true
  line_search = 'none'
  steady_state_detection = true

  petsc_options = '-ksp_type<preonly>'
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'

#  petsc_options = '-pc_svd_monitor'
#  petsc_options_iname = '-pc_type'
#  petsc_options_value = 'svd'
[]

[Outputs]
  exodus = true
  [CSV_L]
    type = CSV
    execute_on = 'FINAL'
    file_base = ./CTF_TEST1/CSV
  [../]
  [VTK]
    type = VTK
    interval = 4
    file_base = ./CTF_TEST1/VTK
  [../]
[]
```
I write these codes in my dg kernel:
```
  std::cout<<"x of normal vector is:"<<_normals[_qp](0)<<std::endl;
  std::cout<<"y of normal vector is:"<<_normals[_qp](1)<<std::endl;
  std::cout<<"z of normal vector is:"<<_normals[_qp](2)<<std::endl;
```
And for my input file, it outputs in the terminal:
```
x of normal vector is:1
y of normal vector is:0
z of normal vector is:0
x of normal vector is:1
y of normal vector is:0
z of normal vector is:0
x of normal vector is:1
y of normal vector is:0
z of normal vector is:0
```
I don't know why.
> Regarding your lateral velocities, the modification you made might be necessary if you have issues with nonlinear convergence. Otherwise, the original may be ok. One question that comes to mind is: do you use this lateral momentum kernel for both of your channels? If so, then you have a problem because one of your channels needs to flip the sign on the conditional.

Yes, I didn't show the code here. The lateral equation will only be solved once. And in the other flow channel, the w is transferred by UO like the way I transfer rhoA, rhouA and rhoEA. And don't worry, I noticed that I should flip the sign.

As you see in the input file, I set the `length` to 500. It's a large number for a pipe. And the result is better than before. I've got the theoretical solution. But this happens at about 100m(which happens at 10m in ctf). I output the mass flwo rate:
```
id,rhouA,x,y,z
500,12.041793375222,0.5,0,0.5
501,12.124408362103,0.5,0,1.5
502,12.205679493002,0.5,0,2.5
503,12.284704437757,0.5,0,3.5
504,12.360901279337,0.5,0,4.5
505,12.433918666798,0.5,0,5.5
506,12.503569841219,0.5,0,6.5
507,12.569784307738,0.5,0,7.5
508,12.632572525886,0.5,0,8.5
509,12.692000164584,0.5,0,9.5
510,12.748169453551,0.5,0,10.5
511,12.801205754223,0.5,0,11.5
512,12.851247960381,0.5,0,12.5
513,12.898441635251,0.5,0,13.5
514,12.942934280111,0.5,0,14.5
515,12.984872016246,0.5,0,15.5
516,13.024397406213,0.5,0,16.5
517,13.06164800882,0.5,0,17.5
518,13.096755502584,0.5,0,18.5
519,13.129845228927,0.5,0,19.5
520,13.161035970839,0.5,0,20.5
521,13.190439966524,0.5,0,21.5
522,13.218163022421,0.5,0,22.5
523,13.244304710319,0.5,0,23.5
524,13.268958625059,0.5,0,24.5
525,13.292212676207,0.5,0,25.5
526,13.314149404741,0.5,0,26.5
527,13.334846262621,0.5,0,27.5
528,13.354375948031,0.5,0,28.5
529,13.372806693046,0.5,0,29.5
530,13.390202520056,0.5,0,30.5
531,13.406623519368,0.5,0,31.5
532,13.422126131421,0.5,0,32.5
533,13.436763342055,0.5,0,33.5
534,13.450584920693,0.5,0,34.5
535,13.463637636174,0.5,0,35.5
536,13.47596544483,0.5,0,36.5
537,13.487609667192,0.5,0,37.5
538,13.498609132374,0.5,0,38.5
539,13.509000393737,0.5,0,39.5
540,13.51881784462,0.5,0,40.5
541,13.5280938577,0.5,0,41.5
542,13.536858918228,0.5,0,42.5
543,13.545141697076,0.5,0,43.5
544,13.552969236855,0.5,0,44.5
545,13.560366986777,0.5,0,45.5
546,13.567358943782,0.5,0,46.5
547,13.57396771458,0.5,0,47.5
548,13.58021461249,0.5,0,48.5
549,13.586119731434,0.5,0,49.5
550,13.591702017823,0.5,0,50.5
551,13.596979343588,0.5,0,51.5
552,13.601968542763,0.5,0,52.5
553,13.606685540142,0.5,0,53.5
554,13.611145342438,0.5,0,54.5
555,13.615362132636,0.5,0,55.5
556,13.619349274823,0.5,0,56.5
557,13.623119385104,0.5,0,57.5
558,13.62668437388,0.5,0,58.5
559,13.630055496122,0.5,0,59.5
560,13.633243384048,0.5,0,60.5
561,13.636258055687,0.5,0,61.5
562,13.639108997776,0.5,0,62.5
563,13.641805167014,0.5,0,63.5
564,13.644355004735,0.5,0,64.5
565,13.646766501485,0.5,0,65.5
566,13.649047208491,0.5,0,66.5
567,13.651204257506,0.5,0,67.5
568,13.653244390193,0.5,0,68.5
569,13.655173960477,0.5,0,69.5
570,13.656998990569,0.5,0,70.5
571,13.658725171581,0.5,0,71.5
572,13.660357880506,0.5,0,72.5
573,13.661902197871,0.5,0,73.5
574,13.663362925233,0.5,0,74.5
575,13.664744599626,0.5,0,75.5
576,13.666051521821,0.5,0,76.5
577,13.667287744499,0.5,0,77.5
578,13.668457100621,0.5,0,78.5
579,13.669563214049,0.5,0,79.5
580,13.670609514522,0.5,0,80.5
581,13.67159925177,0.5,0,81.5
582,13.672535483814,0.5,0,82.5
583,13.673421114039,0.5,0,83.5
584,13.674258874883,0.5,0,84.5
585,13.6750513637,0.5,0,85.5
586,13.675801024692,0.5,0,86.5
587,13.676510178651,0.5,0,87.5
588,13.677181024721,0.5,0,88.5
589,13.677815634051,0.5,0,89.5
590,13.678415964473,0.5,0,90.5
591,13.678983877117,0.5,0,91.5
592,13.679521118856,0.5,0,92.5
593,13.680029362665,0.5,0,93.5
594,13.680510157565,0.5,0,94.5
595,13.680964989727,0.5,0,95.5
596,13.68139526748,0.5,0,96.5
597,13.681802313852,0.5,0,97.5
598,13.682187390752,0.5,0,98.5
599,13.682551670516,0.5,0,99.5
600,13.682896272559,0.5,0,100.5
601,13.683222270307,0.5,0,101.5
602,13.683530684524,0.5,0,102.5
603,13.683822466099,0.5,0,103.5
604,13.684098510831,0.5,0,104.5
605,13.684359663599,0.5,0,105.5
606,13.68460671575,0.5,0,106.5
607,13.684840423047,0.5,0,107.5
608,13.685061507321,0.5,0,108.5
609,13.685270673374,0.5,0,109.5
610,13.685468557763,0.5,0,110.5
611,13.68565575632,0.5,0,111.5
612,13.68583285412,0.5,0,112.5
613,13.686000383973,0.5,0,113.5
614,13.686158877768,0.5,0,114.5
615,13.6863088135,0.5,0,115.5
616,13.686450667461,0.5,0,116.5
617,13.686584871806,0.5,0,117.5
618,13.686711842528,0.5,0,118.5
619,13.686831966343,0.5,0,119.5
620,13.68694561021,0.5,0,120.5
621,13.687053119773,0.5,0,121.5
622,13.68715481933,0.5,0,122.5
623,13.687251037245,0.5,0,123.5
624,13.687342064074,0.5,0,124.5
625,13.687428181323,0.5,0,125.5
626,13.687509651822,0.5,0,126.5
627,13.687586736954,0.5,0,127.5
628,13.687659670319,0.5,0,128.5
629,13.687728673442,0.5,0,129.5
630,13.687793943166,0.5,0,130.5
631,13.687855683302,0.5,0,131.5
632,13.687914098132,0.5,0,132.5
633,13.687969357207,0.5,0,133.5
634,13.688021629004,0.5,0,134.5
```
It's a good news. Does this indicate that I should improve the spatial convergence of the equation? Do you have any ideas about this?

Ethan

**Reply by joshuahansel** [reply]
I'm not sure about the normal vector thing. I vaguely remember something weird about normal vectors and 1D elements (where only the first coordinate is used), so I guess that is what is happening.

> Does this indicate that I should improve the spatial convergence of the equation?

Are you asking if it means that you need to refine your mesh? If so, I'd definitely try a finer mesh to see if you get to the theoretical solution at the theoretical position.
